Add Boleto and OXXO

leonardola opened the pull request:
Fixes #2066

## Changes proposed in this Pull Request:
Add Boleto and OXXO payment methods

## Testing instructions
Follow steps inside the manual test files
[Boleto](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8a70a84d5fe7d976150c9d8858a93bf43f0114a5/tests/manual/stripe-boleto.md)
For boleto the minimum amount is R$5 and maximum R$49999.99. An error must show up when out of this range during checkout

[OXXO](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8a70a84d5fe7d976150c9d8858a93bf43f0114a5/tests/manual/stripe-oxxo.md)
For OXXO the minimum amount is $10 and maximum R$10000. An error must show up when out of this range during checkout

Please test this with and without UPE active. To activate UPE just set _wcstripe_feature_upe_settings to yes on wp_options

**Post merge**

- [x] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
- [x] Check [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2115#issuecomment-969051564)

leonardola committed:
added boleto as a payment method and gateway #2066

leonardola committed:
added boleto on new settings page #2066

leonardola committed:
it is now possible to create an order with boleto #2066

leonardola committed:
added boleto icons #2066

leonardola committed:
updated the boleto icon inside wp-admin #2066

leonardola committed:
added input for CPF, added order note with payment intent id #2066

leonardola committed:
added cnpj support #2066

leonardola committed:
fixed problem that would not allow boleto orders to be set as complete/processing when using upe checkout #2066

leonardola committed:
reverted unecessary change #2066

leonardola committed:
now the cpf/cnpj mask is applied right after the page load #2066

leonardola committed:
added manual test #2066

leonardola committed:
updated comments #2066

leonardola committed:
added input for CPF, added order note with payment intent id #2066

leonardola committed:
added input for CPF, added order note with payment intent id #2066

leonardola committed:
fixed payment method name #2066

leonardola committed:
fixed typo #2066

leonardola committed:
fixed typo #2066

leonardola committed:
Merge remote-tracking branch 'origin/add/2066-add-boleto-and-oxxo' into add/2066-add-boleto-and-oxxo

leonardola committed:
added OXXO Payment Method #2066

leonardola committed:
it is now possible to checkout using OXXO #2066

leonardola committed:
fixed some typos and disabled reusable on boleto and oxxo #2066

leonardola committed:
fixed problem that would not allow to select OXXO on upe checkout #2066

leonardola committed:
added OXXO manual test #2066

leonardola committed:
added changelog entry #2066

leonardola committed:
fixed tests after adding Boleto and OXXO #2066

leonardola committed:
Merge branch 'develop' into add/2066-add-boleto-and-oxxo

# Conflicts:
#	changelog.txt

leonardola committed:
Removed payment method description from OXXO and Boleto #2066

leonardola committed:
now the refund via stripe does not show on admin when order was paid with Boleto or OXXO #2066

leonardola committed:
added comments and some white spaces #2066

leonardola committed:
added amount limits on OXXO #2066

leonardola committed:
added some tests to validate  #2066

a-danae reviewed (commented):
Nice stuff 🙂  dropping here some thoughts about Boleto:

1 - There's a gotcha that wasn't obvious to me: The country of the [Stripe account](https://dashboard.stripe.com/settings/account) you're using must be set to Brazil for it to work. If the Stripe account country isn't set to Brazil, you'll get this error during checkout, even when the store and the shopper locations are set to Brazil in WC.

 Because there's nothing the shopper could do about this, could we let the merchant know about this at some point when enabling Boleto in WC's settings? Not sure if we get this information from Stripe at some point, but if so, maybe we could check this directly before enabling Boleto.

The WC log says `Create payment intent error: The payment method "boleto" is not supported for payments settling in country "US"`. And here's a screenshot of the error I get during checkout.
![image](https://user-images.githubusercontent.com/41606954/140621941-91eb0674-bdee-463a-b310-f5334a782361.png)

2 - Go to checkout and select Boleto. Add "123" as the shopper's email address and click on "Place order". Notice nothing happens. When you open the browser's console, it says "Your email address is invalid". We should display some notice to the shopper so they can fix their error in the form.

This is true for other error notices. For example, when using a five-digit ZIP code, this gets logged in the browser console with no messages for the shopper: `Invalid postal code for Brazil. Must be an eight-digit code registered by Correios`. Rather than just display it right away, it'd be nice if we could make it translatable. Probably what we're doing [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-helper.php#L193) for other errors?

For reference, this notice shows up when you do the same with the "Credit / debit card" method.
![image](https://user-images.githubusercontent.com/41606954/140622922-ba75ae5e-13f0-4f04-9245-6287bd721229.png)

3 - Is it expected that the shopper could get the modal with the voucher when the form still has errors? Video of what I mean https://d.pr/v/3Zg8gX . The payment intent remains "Incomplete" on Stripe, so all is good on that end. I just find it confusing as a shopper. It also looks like if you change any of the details in the form, these details in the WC order will differ from what's on the Payment intent in Stripe.
![image](https://user-images.githubusercontent.com/41606954/140624062-a10878eb-bd1f-4c7f-aefd-f63d8a7ac137.png)


Btw, are Boleto and Oxxo functionalities entangled together? Wondering whether it'd be convenient to have a separate PR for each.

leonardola committed:
added error message area for boleto #2066

leonardola committed:
added error message area for OXXO

leonardola committed:
added error message for invalid postal code #2066

leonardola committed:
added messages for too low and too high totals and invalid country codes

leonardola committed:
added validation for required fields before showing boleto or OXXO voucher

leonardola committed:
added messages for required fields

leonardola committed:
added message for required cpf/cnpj

leonardola committed:
added info about setting country on stripe dashboard for boleto #2066

leonardola committed:
added info about setting country on stripe dashboard for OXXO #2066

leonardola committed:
now boleto is shown only on success page not during checkout #2066

leonardola committed:
not it places the order then shows the boleto before the success page #2066

leonardola committed:
added description with order id when creating a stripe payment intent #2066

leonardola committed:
fixed problem that would not allow to set a boleto order to processing/complete #2066

leonardola committed:
now the order is placed before showing the oxxo voucher #2066

leonardola commented:
@a-danae 

> There's a gotcha that wasn't obvious to me: The country of the Stripe account you're using must be set to Brazil for it to work. If the Stripe account country isn't set to Brazil, you'll get this error during checkout, even when the store and the shopper locations are set to Brazil in WC.

Oops I added that to the test flow now

> We should display some notice to the shopper so they can fix their error in the form

I changed how the payment is processed. Now it tries to place the order(so the default checks run) and only show the boleto/voucher if everything went ok. If anything fails during checkout the boleto/voucher is not shown and the shopper can fix the errors and try again. But if an error happens when trying to show the boleto/voucher the shopper will have to order again as the order was already placed and the cart emptied. This is how the upe worked before I did any changes so at least now the behavior is the same in both cases and many more check execute before trying to show the boleto/voucher.

>Is it expected that the shopper could get the modal with the voucher when the form still has errors? Video of what I mean https://d.pr/v/3Zg8gX . The payment intent remains "Incomplete" on Stripe, so all is good on that end. I just find it confusing as a shopper. It also looks like if you change any of the details in the form, these details in the WC order will differ from what's on the Payment intent in Stripe.

With the changes above the checkout errors will show before the boleto/voucher is shown.

>Btw, are Boleto and OXXO functionalities entangled together? Wondering whether it'd be convenient to have a separate PR for each.

They are not entangled together but they share some code so It would take some copy pasting to separate them into multiple PRs and fixing some conflicts after. But I can separate them if that is necessary

leonardola committed:
added comments and checkout description message configuration on OXXO #2066

leonardola committed:
now customers can only place order with boleto or OXXO if correct country is selected. Fixed bug that would not allow to place orders when using only boleto or only OXXO  #2066

leonardola committed:
added link to giropay guide page #2066

ricardo commented:
> Because there's nothing the shopper could do about this, could we let the merchant know about this at some point when enabling Boleto in WC's settings? Not sure if we get this information from Stripe at some point, but if so, maybe we could check this directly before enabling Boleto.

@a-danae, good suggestion.

I think this can be fixed with https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2143, which shows a "pending" indication or simply hides boleto if it's not supported for that Stripe account.

LMK if you all have any thoughts or need help with the review as I have the PR reviewer hat this week for Fractal, and I'm also familiar with boleto 👍

leonardola commented:
@ricardo #2143 will only apply to upe methods and the new settings page right? Do I need to do anything to add support for boleto and OXXO or your solution will apply to them too?

leonardola committed:
Merge branch 'develop' into add/2066-add-boleto-and-oxxo

# Conflicts:
#	changelog.txt
#	includes/class-wc-gateway-stripe.php

leonardola committed:
added changes from develop branch #2066

ricardo commented:
@leonardola The new settings will be the default after the next release, so we don't need to worry about updating the old settings. I should mention we need to test this PR with the new settings feature flag enabled: `_wcstripe_feature_upe_settings` option set to `yes`.

#2143 should display the pending activation notice to the non-UPE settings page too, but it will not hide non-UPE payment methods when they're not supported, e.g Boleto for US merchants - We don't do that to preserve current behavior.

> Do I need to do anything to add support for boleto and OXXO

I believe the only thing you need to do is add the Stripe payment method ID [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2143/files#diff-49dc330736a1b24c7a8eb18da77a0402aee77e0d8f8a5f03b25bdcb30a8492f9R5) for Boleto and OXXO. But we should probably create a separate issue to test Boleto + OXXO with these flows that were just merged or are not quite merged yet - from what I remember:
- https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2143
- https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2130

And an extra issue to add these new APMs to the documentation: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1763

Since you're the one working on it, would you like to create these issues once this PR gets merged?

leonardola commented:
@ricardo you can leave it to me. I'll validate everything after both PRs are merged and create issues accordingly.

a-danae commented:
Thanks for the updates @leonardola! 🙂 

From my previous comment:

1 - Thanks for adding it to the testing instructions! 
Reading the comments above, then we'd need the PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2143 for either showing a notice or hiding Boleto if the Stripe account, right? Because that PR is blocked, I think it'd be good to have a separate issue for updating Boleto. I don't see that as a blocker for this PR, but maybe we could add some copy so the merchant can learn about this limitation. What do you think?

2 and 3 - Sweet, now these work as expected 🎉 

> But if an error happens when trying to show the boleto/voucher the shopper will have to order again as the order was already placed and the cart emptied.

Do you have steps to make this happen? Asking for testing. If the order is marked as "Failed" or "Pending payment", the shopper could easily get the voucher again via the Pay for order page. Although I'm not sure if that's something we could control in this scenario.

Regarding having separate PRs for OXXO and Boleto,
> But I can separate them if that is necessary

I think we're good, it'd do more harm than good at this point since there's some overlapping code. Thanks!

Things seem to work fine for Boleto 🙂  going with OXXO next.

a-danae commented on includes/payment-methods/class-wc-stripe-upe-payment-method.php:
```diff
@@ -60,6 +60,13 @@ abstract class WC_Stripe_UPE_Payment_Method {
+	 *
+	 * @var array
+	 */
+	public $can_refund = true;
```
@leonardola this works fine, I was just wondering about following the pattern for other properties: making it protected and having a public getter to retrieve it. What do you think?

a-danae commented on tests/manual/stripe-boleto.md:
```diff
@@ -0,0 +1,62 @@
+## Stripe Alipay
```
"Boleto" instead of "Alipay" probably? 😄  (same for OXXO)

a-danae reviewed (dismissed):
Things work fine, as I mentioned in the comment above. 

There's one detail we should probably check out: Boleto doesn't seem to be working on the Pay for order page https://d.pr/v/ecExmw 

Thanks again for the updates 🙂

leonardola committed:
fixed test description

leonardola commented:
About fails after placing order
> Do you have steps to make this happen? Asking for testing. If the order is marked as "Failed" or "Pending payment", the shopper could easily get the voucher again via the Pay for order page. Although I'm not sure if that's something we could control in this scenario.

If you add an invalid CPF for example: 375.554.650-77. An error will show and if you try to place the order again it will throw a session expired error. That happens because the cart is emptied before stripe checks for the cpf. Maybe I could add some info on the message about reloading the page or having to reorder. Something like "Invalid Tax Id, please try again with a valid tax id. Your cart was emptied please reload the page and order again"

leonardola committed:
added function to validate if payment can be refunded via stripe #2066

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -214,6 +215,10 @@ public static function get_localized_messages() {
 					: __( 'Unable to process this payment, please try again or use alternative method.', 'woocommerce-gateway-stripe' ),
+				'amount_too_large'         => __( 'The order total is too high for this payment method', 'woocommerce-gateway-stripe' ),
+				'amount_too_small'         => __( 'The order total is too low for this payment method', 'woocommerce-gateway-stripe' ),
+				'country_code_invalid'     => __( 'Invalid country code, please try again with a valid country code', 'woocommerce-gateway-stripe' ),
```
@leonardola couple of questions here:

1. With UPE, when selecting a country and state that's not from Brazil, this message shows up. I think it's not immediately clear what the problem is for a shopper, but I don't see a specific code returned by Stripe to handle this like other error codes. Do you think we could improve it?
![image](https://user-images.githubusercontent.com/41606954/142065101-c5c38bc7-bcfe-4e36-a51c-51e34305f8f3.png)
 
2. Are we actually using this string in the checkout when that error is returned by stripe? With UPE enabled, when I select a billing country that's not Brazil (and add "AC" in the state to skip the validation I mentioned before), I see the message returned by the request instead of "Invalid country code, please try again with a valid country code". Even though the message returned by Stripe could be useful, it'd be good to have a translatable string.
![image](https://user-images.githubusercontent.com/41606954/142059206-b8e64c38-fbc8-4ace-b20b-577ee234821a.png)

a-danae commented on includes/payment-methods/class-wc-stripe-upe-payment-method-boleto.php:
```diff
@@ -0,0 +1,33 @@
+ * Boleto Payment Method class extending UPE base class
+ */
+class WC_Stripe_UPE_Payment_Method_Boleto extends WC_Stripe_UPE_Payment_Method {
+
```
@leonardola  Not related to this file but to Boleto UPE. When having an order with an amount that's not within Boleto's limits, say R$4, creating an intent fails, and the iframe for paying with a card isn't loaded. I think the shopper should still be able to pay with a card in this scenario. Is this something we could adjust?
Video with the behavior https://d.pr/v/XB6Wgw

ricardo replied:
Chiming in again as I read the notification 😅. Probably related: #1769. Not sure if this is easy to fix 🤔

leonardola replied:
Maybe we should tackle that in another PR?

a-danae replied:
Glad to see you around, tbh 😄 

Gotcha. Since this seems to be a thing at the root of UPE, I think we could mention this in that issue and handle it there. I'm dropping a message on that issue and marking this conversation as resolved. Thanks Ricardo!

a-danae replied:
@leonardola , I missed your comment when I was typing mine, sorry. Sounds good, yup. Do you want to have a separate issue for this case? Glad to create it. I was thinking of simply mentioning this on #1769 for when someone works on that.

leonardola replied:
I already mentioned it there. So we should be good to go

leonardola committed:
added translation to error messages during check out with UPE #2066

leonardola commented:
> There's one detail we should probably check out: Boleto doesn't seem to be working on the Pay for order page https://d.pr/v/ecExmw

~~I've tested with credit card and it also does not work. This is happening because the UPE gateway does not seem prepared to accept payments on the pay order page. Maybe we could create a new issue to fix this. I think we can ignore when paying without UPE as it will be removed on next release as @ricardo said. I am still waiting on an answer from Fractal to be 100% sure about it~~

The problem is that this option is not selected by default on UPE
<img width="352" alt="Screen Shot 2021-11-17 at 7 07 16 PM" src="https://user-images.githubusercontent.com/6342964/142291401-922a5fb9-2882-4eb5-be2f-fc6ffe7e0163.png">
After selecting it and placing the order with boleto it worked. I'll take another look without UPE

> With UPE, when selecting a country and state that's not from Brazil, this message shows up. I think it's not immediately clear what the problem is for a shopper, but I don't see a specific code returned by Stripe to handle this like other error codes. Do you think we could improve it?

I've added another validation inside the process_payment function to create a better error message



> Are we actually using this string in the checkout when that error is returned by stripe? With UPE enabled, when I select a billing country that's not Brazil (and add "AC" in the state to skip the validation I mentioned before), I see the message returned by the request instead of "Invalid country code, please try again with a valid country code". Even though the message returned by Stripe could be useful, it'd be good to have a translatable string.

Upe was not using translations for errors when placing order. I've added it now but keeping the previous behavior when there is no translation

leonardola committed:
Merge branch 'develop' into add/2066-add-boleto-and-oxxo

# Conflicts:
#	changelog.txt

ricardo commented:
> I think we can ignore when paying without UPE as it will be removed on next release as @ricardo said.

@leonardola I said the new settings will be the default, but we will still allow UPE to be disabled.

a-danae reviewed (commented):
> I've added another validation inside the process_payment function to create a better error message

Looks great!

> Upe was not using translations for errors when placing order. I've added it now but keeping the previous behavior when there is no translation

Sweet, sounds good to me.

> After selecting it and placing the order with boleto it worked. I'll take another look without UPE

The behavior I referred to is actually without UPE. Should have clarified, sorry. Things work fine with UPE enabled, as long as that checkbox you mentioned is checked ~(I'm creating an issue for it in the morning)~.

> If you add an invalid CPF for example: 375.554.650-77. An error will show and if you try to place the order again it will throw a session expired error. That happens because the cart is emptied before stripe checks for the cpf. Maybe I could add some info on the message about reloading the page or having to reorder. Something like "Invalid Tax Id, please try again with a valid tax id. Your cart was emptied please reload the page and order again"

Would it be possible to prevent emptying the card until the validation passes? Thinking that, as a shopper, it'd be frustrating to have to re-order things if you had a typo in the CPF.

These last two points _might_ be blockers in my opinion, but glad to have them as separate issues if either of you thinks they're not critical. Other than that, that'd be it from on the functional side of Boleto 🙂 

Thanks @leonardola!

leonardola committed:
now if an error happens before the customer can see the boleto it is possible to try again without having to add the products to the cart again #2066

leonardola committed:
now it is possible to use boleto on Pay Order Page #2066

a-danae commented on tests/manual/stripe-oxxo.md:
```diff
@@ -0,0 +1,54 @@
+
+#### Set currency to Mexican Peso ($)
+
+#### Set country to Mexico
```
Looks like changing the store country isn't required for OXXO nor Boleto? They both work for me having the United States as the store base.

That aside, what do you think about mentioning testing both the UPE and non-UPE versions in these instructions?

a-danae reviewed (commented):
I'm done with testing now 🙂 leaving here a recap of what we mentioned before and some other behaviors I noticed. Great work on these @leonardola!

**Boleto - UPE and non-UPE - Checkout (mentioned before)**

The cart gets emptied when entering an invalid CFP, like “375.554.650-77”, as Leonardo [mentioned](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2115#issuecomment-970581961). I think that having to re-order because you had a typo in the CFP could be frustrating as a shopper, and could look like a bug.

This is the only behavior I think it'd be good to update before merging. Glad to have it as a separate issue if either of you finds it better.

----

The ones below are behaviors that I think could be improved. Nothing I’d call a blocker, and maybe we wouldn’t want to address some of them at all, but leaving them here for reference and to discuss.

**OXXO and Boleto - Non UPE - Pay for order (mentioned above)**

Can’t pay for the order, even when all details are filled up. Paying with a card does work https://d.pr/v/KVD3C7 

**OXXO and Boleto - Non UPE - Checkout block**

They don’t show up in the available payment methods.

**OXXO and Boleto - UPE - Pay for order**

The order remains as “Failed” after further payment attempts if you didn’t select “Use a new payment method” when trying to pay the first time https://d.pr/v/g9ayjb This won’t happen anymore when [this issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2163) gets solved.

**OXXO and Boleto - UPE - Pay for order**

The order remains as “On hold” when the Stripe validation fails. For example, with Boleto, using “375.554.650-77” as the CFP and then trying again with “000.000.000-00”. Or with OXXO adding a single name in the fields and then filling the field with the correct details https://d.pr/v/hZfwqh Even after waiting, the payment intent remains as “Incomplete” and the order remains on hold

**OXXO and Boleto - UPE - Edit order**

The payment method that was used isn’t displayed in the Edit order page, but we do display it when paying with at least a Debit / Credit card and Giropay, which are the ones I tested 
![image](https://user-images.githubusercontent.com/41606954/142544921-1b8769f7-c5b6-46b7-bac1-a04cecf89743.png)


**OXXO and Boleto - Checkout (mentioned above)**

An error message shows up when trying to check out if the Stripe account linked to the store is in a non-supported country. Because there’s nothing the shopper could do about this, it’d be good to make the merchant aware of this instead at some point. Ricardo [mentioned](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2115#issuecomment-966613142) that this behavior might be fixed by https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2143, so this point might not be relevant after integrating that change.


-----

**A quick overview of what I tested with Boleto and OXXO, both as UPE and non-UPE:**

- Selecting a non-supported country in the Stripe account
- Selecting a non-supported currency on the site
- Checking out with a non-supported country as the client’s billing country
- Checking out with required billing fields being empty
- Checking out as a guest and as a logged-in user
- Triggering all different payment outcome scenarios. OXXO, Boleto
- Trying to make a refund of an order paid with OXXO or Boleto
- Using the Pay for order page to pay for an order as a guest and as a logged-in user
- Using the Pay for order page to pay for orders with and without the billing details filled up


It'd be nice if you could check it out as well, @ricardo in case I missed something or in case I experienced some behavior that isn't really happening.

Thanks!

leonardola committed:
Created a class to make implementing voucher type payments easier #2066

leonardola committed:
now oxxo works on pay order. Refactor on class-wc-gateway-stripe-oxxo to extend class-wc-gateway-stripe-voucher #2066

leonardola committed:
hidden the refund via stripe button #2066

leonardola committed:
fixed bug that would not allow boleto and oxxo to show up when UPE is active #2066

leonardola committed:
added instruction to execute manual test again with UPE on #2066

leonardola committed:
removed instruction to set country on manual test  #2066

leonardola committed:
fixed problem that would make orders stay on failed state after successfull payment with oxxo or boleto on Pay Order page #2066

ricardo commented:
> It'd be nice if you could check it out as well, @ricardo

The settings page for Boleto appears to be blank for me (new settings feature flag set to `yes`).

<img width="1153" alt="Boleto settings" src="https://user-images.githubusercontent.com/5509901/142702914-1cf272fc-d92b-4aa3-881d-c19dd8ee5e84.png">

It should instead be similar to the SEPA settings and have a section about webhooks.

Also, how about we add a parameter to this page to set the number of days for the voucher to expire? It could be a further improvement for a separate PR perhaps. We'd need to save that as a setting and then pass as a parameter to the payment intent.

https://stripe.com/docs/payments/boleto/accept-a-payment?platform=web#additional-payment-method-options

ricardo commented on assets/js/stripe.js:
```diff
@@ -264,6 +264,20 @@ jQuery( function( $ ) {
 			wc_stripe_form.maybeConfirmIntent();
+
+			//Mask CPF/CNPJ field when using Boleto
+			$( document ).on( 'change', '.woocommerce-checkout-payment', function() {
+				var TaxIdMaskBehavior = function (val) {
+						return val.replace(/\D/g, '').length >= 12 ? '00.000.000/0000-00' : '000.000.000-009';
+					},
+					spOptions = {
+						onKeyPress: function(val, e, field, options) {
+							field.mask(TaxIdMaskBehavior.apply({}, arguments), options);
+						}
+					};
+
+				$( '#stripe_boleto_tax_id' ).mask( TaxIdMaskBehavior, spOptions );
+			});
```
I can't seem to add a CNPJ here. The mask doesn't let me do it.

Also, we need to normalize this code with the WP coding standards. More specifically, we're missing spaces before/after parenthesis.

It's unfortunate that the Stripe elements API doesn't provide a CPF tax ID form field, so we have to write our own and include the mask dependency.

leonardola replied:
It seems like if you type the CNPJ it works but if you try to paste it, it does not work. I'll take a look at it

leonardola replied:
fixed

ricardo replied:
Let me know when you push the change, so I can re-review with the applied fixes.

It also helps a lot when you reference the hash on the "Fixed/resolve conversation" reply.

ricardo commented on client/payment-methods-map.js:
```diff
@@ -98,4 +100,26 @@ export default {
+		id: 'boleto',
+		label: __( 'Boleto', 'woocommerce-gateway-stripe' ),
+		description: __(
+			'Boleto is one of the most popular payment method in Brazil',
```
Where does this description come from? Shouldn't it be "payment method**s**"?

I think the other descriptions here are based on Stripe docs: https://stripe.com/docs/payments/boleto

leonardola replied:
What is your suggestion for a description here? 

"Boleto is an official (regulated by the Central Bank of Brazil) payment method in Brazil." Also doesn't seem very informative. And coping the whole thing seems a bit too long.

Maybe "Boleto is an official payment method in Brazil. Customers receive a voucher that can be paid on banks, ATMs, or online bank portals. Payment confirmation takes 1 business day and funds are available after 2  business days after confirmation" would be better?

ricardo replied:
I agree the first sentence from the Stripe docs for Boleto doesn't sound good.

I noticed most of the other descriptions combine the first two sentences from the Stripe docs, but aren't necessarily too descriptive. I don't think we need to inform how many business days the funds are available, etc. so I think we can cut that out:

> Boleto is an official payment method in Brazil. Customers receive a voucher that can be paid at authorized agencies or banks, ATMs, or online bank portals.

That is more or less the size of the P24 description.

Also, the current OXXO description doesn't look right to me. How about this?

> OXXO is a Mexican chain of convenience stores that allows customers to pay bills and online purchases in-store with cash.

ricardo commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -543,7 +545,7 @@ public function process_response( $response, $order ) {
 				throw new WC_Stripe_Exception( print_r( $response, true ), $localized_message );
 			}
-		} else {
+		} elseif ( 'requires_payment_method' !== $response->status ) {
```
Can you provide some context as to why this change is needed?

leonardola replied:
If an error happens before the customer can see the boleto(like wrong cpf), stripe sends a notification that would end up putting the order on hold and reducing the stock levels. When using OXXO and Boleto the status is requires_payment_method. Other payment methods use other status so that seem the safest way to check. I'll take a deeper look at it to try to find another solution.

leonardola replied:
I changed how the errors are handled. Now if the chosen payment method is boleto or oxxo the `update_failed_order_ajax` ajax endpoint is not called

ricardo replied:
So the above change is no longer necessary? Please LMK when you have pushed the change.

ricardo commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1365,6 +1367,8 @@ public function update_existing_intent( $intent, $order, $prepared_source ) {
 			$request['customer'] = $prepared_source->customer;
 		}
 
+		$request['payment_method_types'] = [ 'card' ];
```
Also, why is this needed? I'm afraid it may change current behavior.

leonardola replied:
This function is only being used by stripe credit card payment. Maybe it should be inside `WC_Gateway_Stripe` instead of `WC_Stripe_Payment_Gateway`.

This change is necessary in case a customer tries to pay with Boleto or OXXO but it fails,  then the customer tries to checkout with credit card.

ricardo reviewed (commented):
I began to review this, but won't be able to finish now as I have some janitor priorities.

It would be good to have someone else from Fractal to review this PR as it contains some refactorings that may need some checking.

cc @reykjalin If you have the bandwidth/are ok with it.

leonardola committed:
fixed description #2066

leonardola committed:
fixed code style #2066

leonardola committed:
fixed problem that would not allow to paste a CNPJ on boleto #2066

leonardola committed:
now CPF/CNPJ mask is applied on pay order page #2066

leonardola committed:
fixed settings page when wp_option _wcstripe_feature_upe_settings is enabled for OXXO and Boleto #2066

leonardola committed:
changed how boleto and oxxo are treated on checkout/pay order failure #2066

leonardola committed:
fixed bug that would not reduce stock for boleto or OXXO on UPE. Removed unecesary code for non UPE after changes #2066

leonardola committed:
fixed order status after error when using invalid credit card then boleto on UPE #2066

leonardola committed:
fixed bug that would allow refund via stripe on Boleto and Oxxo when using blocks checkout #2066

leonardola committed:
fixed bug that would make block checkout reuse orders that were waiting payments on Boleto and OXXO #2066

leonardola committed:
fixed bug that would not allow orders to go into on-hold and into processing when using boleto/oxxo on classic checkout #2066

leonardola committed:
added flow to put orders on failed state when boleto or oxxo voucher expires #2066

leonardola committed:
fixed problem that would allow refund via stripe on OXXO when paying from pay order page #2066

leonardola committed:
fixed bug that would allow wrong country to place order using OXXO #2066

leonardola committed:
Merge branch 'develop' into add/2066-add-boleto-and-oxxo

# Conflicts:
#	client/payment-methods-map.js

leonardola committed:
added description on showError function #2066

leonardola committed:
updated boleto and OXXO descriptions

leonardola committed:
added dockblock on is_allowed_on_coutry function #2066

leonardola committed:
fixed test error after adding payment intent id to orders #2066

leonardola commented:
## Bugs found
### Non UPE
- [x] Do not empty cart before success page
- [x] Make pay order page work
- [x] Only Manual Refund should show on Edit Order
- [x] Only Manual Refund should show on Edit Order after using Pay Order page
- [x] Allow use of payment method alone
- [ ] Blocks checkout. (#2117 might fix)
- [x] Reduce stock correctly
- [x] Allow to paste CNPJ during checkout and pay order for boleto
- [x] Apply cpf mask on on Pay Order page for boleto

### UPE
- [x] Blocks checkout should work
- [x] Only Manual Refund should show on Edit Order
- [x] Only Manual Refund should show on Edit Order after using Pay Order page
- [x] Add instructions for UPE on manual tests
- [x] Order remains on hold after failed trye and then successfull try inside pay for order
- [ ] Payment Method name does not show on edit order (also happens with giropay UPE. Issue will be created)
- [x] Reduce stock correctly
- [ ] Show correct payment method on order edit page(also happens with giropay UPE. Issue will be created)

### With _wcstripe_feature_upe_settings active
- [x] With UPE Disabled settings are not working


### Both UPE and non UPE
- [ ] Checkout error when wrong country is set on stripe dashboard. (#2143 might fix. May need extra changes after both PR as merged)
- [x] Remove country change from manual test

## Tests

**This are the stages which an order should go through when using a voucher payment method**
- Click on Pay order. Order should be created on admin and status should be Pending Payment
- Modal should pop-up
- Stripe will send a notification of type payment_intent.requires_action. The order status will be changed to on-hold. Stock will be decreased
- Closing the modal will redirect to success page
- After 4 minutes stripe will send a notification of type payment_intent.succeeded. Order status will be changed to processing

On each test make sure the payment was approved on Stripe and the order went to processing.
Execute the tests below with the following checkouts:

- with UPE
- with UPE blocks checkout
- with non UPE


### Boleto
- Checkout with all correct parameters
- Checkout with wrong cpf then correct one
- Checkout with wrong cpf then credit card
- Checkout with invalid credit card then boleto
- Checkout with invalid card then valid card
- Checkout with wrong cpf then use boleto on pay order page
- Checkout with wrong cpf then use credit card on pay order page
- Checkout with invalid card then use boleto on pay order page
- Checkout with invalid card then use card on pay order page
- Try to checkout with Invalid country
- Try to checkout with 5 digit zipcode
- Checkout with only boleto enabled
- Only manual refund should show on edit page when paid from checkout
- Only manual refund should show on edit page when paid from pay order
- Correct payment method on order edit page from checkout (does not work with upe on classic checkout)
- Correct payment method on order edit page from pay order (does not work with upe on classic checkout)
- Transaction Id shows up on edit page after boleto is paid
- Enable _wcstripe_feature_upe_settings but do not enable UPE. Open boleto settings
- Stock is only decreased after payment_intent.requires_action notification on checkout
- Stock is only decreased after payment_intent.requires_action notification on pay order

### OXXO
- Checkout with all correct parameters
- Checkout with invalid credit card then oxxo
- Checkout with invalid card then valid card
- Checkout with invalid card then use oxxo on pay order page
- Checkout with invalid card then use card on pay order page
- Try to checkout with Invalid country
- Checkout with only oxxo enabled
- Only manual refund should show on edit page when paid from checkout
- Only manual refund should show on edit page when paid from pay order
- Correct payment method on order edit page from checkout (does not work with upe on classic checkout)
- Correct payment method on order edit page from pay order (does not work with upe on classic checkout)
- Transaction Id shows up on edit page after boleto is paid
- Enable _wcstripe_feature_upe_settings but do not enable UPE. Open OXXO settings
- Stock is only decreased after mpayment_intent.requires_action notification on checkout
- Stock is only decreased after payment_intent.requires_action notification pay order


## Out of scope issues found
- When using blocks checkout if you click on place order and it fails then you update any field the outdated data is sent to stripe #2195 
- includes/class-wc-gateway-stripe.php on line 436 error shows up when using invalid card on pay order non upe #2196
- duplicate error messages on blocks checkout #2093
- Add settings for expire date on boleto https://stripe.com/docs/payments/boleto/accept-a-payment?platform=web#additional-payment-method-options #2198
- Remove html from translatable strings on settings pages on [all payment methods(https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/admin/stripe-bancontact-settings.php#L14) #2199
- Customer has to select the state every time on blocks checkout [#5259](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5259)


@reykjalin This PR is ready for a new review. Hope this time I didn't miss anything. 
@ricardo and @a-danae in case you want to give it a try again

reykjalin commented on assets/js/jquery.mask.min.js:
```diff
@@ -0,0 +1,19 @@
+// jQuery Mask Plugin v1.14.16
```
Just to be sure — is this plugin included with WordPress? Can we fall-back to a version included with WP by including it in the dependencies when we register the JS?

reykjalin commented on includes/admin/stripe-oxxo-settings.php:
```diff
@@ -0,0 +1,49 @@
+			'type'        => 'title',
+		],
+		'guide'       => [
+			'description' => __( '<a href="https://stripe.com/payments/payment-methods-guide#oxxo" target="_blank">Payment Method Guide</a>', 'woocommerce-gateway-stripe' ),
```
I think it's generally very bad practice to include HTML tags in translatable strings. We should really use a `sprintf` approach like we do in other places in the codebase, e.g.:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/fe0d2951ebdc236ebc18302ab161b15f3451b48f/includes/class-wc-stripe-apple-pay-registration.php#L396-L401

and

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/fe0d2951ebdc236ebc18302ab161b15f3451b48f/includes/admin/stripe-settings.php#L120-L128

reykjalin commented on includes/class-wc-stripe-intent-controller.php:
```diff
@@ -404,14 +404,16 @@ public function update_payment_intent( $payment_intent_id = '', $order_id = null
+				'amount'      => WC_Stripe_Helper::get_stripe_amount( $amount, strtolower( $currency ) ),
+				'currency'    => strtolower( $currency ),
+				'metadata'    => $gateway->get_metadata_from_order( $order ),
+				'description' => __( 'stripe - Order', 'woocommerce-gateway-stripe' ) . ' ' . $order->get_id(),
```
Should the "stripe" be capitalized here? 🤔

reykjalin commented on includes/class-wc-stripe-webhook-handler.php:
```diff
@@ -801,10 +801,17 @@ public function process_payment_intent_success( $notification ) {
 		}
 
+		$is_voucher_payment = in_array( $order->get_meta( '_stripe_upe_payment_type' ), [ 'boleto', 'oxxo' ] );
+		$allowed_status     = [ 'pending', 'failed' ];
```
Let's rename this variable to use the plural form `$allowed_statuses` 🙂

reykjalin replied:
Also, would it make more sense to use the `wc_stripe_allowed_payment_processing_statuses` filter in the Boleto and OXXO classes instead of using an if-statement here? 🤔

reykjalin commented on includes/payment-methods/class-wc-gateway-stripe-boleto.php:
```diff
@@ -0,0 +1,138 @@
+			throw new WC_Stripe_Exception( sprintf( __( 'Sorry, the minimum allowed order total is %1$s to use this payment method.', 'woocommerce-gateway-stripe' ), wc_price( 5.00 ) ) );
+		}
+
+		if ( $amount > 49999.99 ) {
```
I think it would be better to use `elseif` here because this is a single if-statement in terms of semantics. There's no way both of these conditions can be true.

reykjalin commented on includes/payment-methods/class-wc-gateway-stripe-oxxo.php:
```diff
@@ -0,0 +1,111 @@
+		if ( $amount > 10000.00 ) {
+			/* translators: 1) amount (including currency symbol) */
+			throw new WC_Stripe_Exception( sprintf( __( 'Sorry, the maximum allowed order total is %1$s to use this payment method.', 'woocommerce-gateway-stripe' ), wc_price( 10000.00 ) ) );
+		}
```
Same as with Boleto; let's use `elseif` here.

reykjalin commented on includes/payment-methods/class-wc-gateway-stripe-voucher.php:
```diff
@@ -0,0 +1,389 @@
+	 * @since 5.8.0
+	 */
+	protected function validate_amount_limits( $amount ) {
+	}
```
Since this function is intended to be overridden should this function (and the class) be abstract?

reykjalin commented on tests/phpunit/test-class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -145,13 +145,13 @@ private function get_order_details( $order ) {
+			'site_url'          => 'http://example.org',
+			'order_id'          => $order_id,
+			'order_key'         => $order_key,
+			'payment_type'      => 'single',
```
I don't think the extra spaces here are needed? 🤔

reykjalin reviewed (changes requested):
Non-UPE checkout doesn't work for me at all with Boleto or OXXO. My first instinct is that there's something not quite right in the JS? The reason that's my first idea is that the formatting isn't applied at all to the test tax IDs I enter in the CPF/CNPJ field when using Boleto.

I can click "place order" and I'm taken straight to the "order received" page; no pop-up is displayed.

It's like the JS doesn't have the opportunity to confirm the payment via Stripe's JS and that's why the modal never pops up.

The payments are set to "Incomplete" in the Stripe dashboard.

***

Everything seems to work fine with the UPE element 🎉

leonardola committed:
fixed translatable strings with html #2066

leonardola committed:
Fixed case of Stripe #2066

leonardola committed:
now the on-hold status is added to the allowed status array via a filter #2066

leonardola committed:
changed to elseif as both conditions at the same time were impossible #2066

leonardola committed:
fixed typo #2066

leonardola committed:
transformed WC_Stripe_Payment_Gateway_Voucher into an abstract class #2066

leonardola committed:
removed empty spaces #2066

leonardola commented:
> I can click "place order" and I'm taken straight to the "order received" page; no pop-up is displayed.
It's like the JS doesn't have the opportunity to confirm the payment via Stripe's JS and that's why the modal never pops up.

Are you setting the `SCRIPT_DEBUG` constant to true?  If not then the minified js version will be used which is out of date. Should I update the minified version during PR or it will be updated later?

reykjalin commented:
> Are you setting the `SCRIPT_DEBUG` constant to true? If not then the minified js version will be used which is out of date.

Yep! But I'll try minifying the JS and see if that works.

**EDIT:** @leonardola it still doesn't work after I minified the assets 🤔 

> Should I update the minified version during PR or it will be updated later?

Yes, please push the minified JS 🙂 
It is generated automatically when we run `woorelease`, but we try to keep the minified assets in the repo anyway 👍

leonardola committed:
fixed problem that would remove jquery mask during npm run build #2066

leonardola committed:
updated minified files #2066

leonardola committed:
Merge branch 'develop' into add/2066-add-boleto-and-oxxo

diegocurbelo committed:
Remove [ and ] from the status check name (#2194)

* Replace [] with () for the status check names

brianyu28 committed:
Add REST API to create customer for an order (#2159)

* Add API endpoint to create Stripe customer

* Add tests for customer creation

* Standardize return values for WP_Error cases

* Fix bug for customer id with empty user

leonardola committed:
fixed bug that would not allow orders using boleto/oxxo to go to processing on upe checkout and bug that would not show boleto as the payment method on edit order when using upe #2066

leonardola committed:
fixed bug that would not allow orders using boleto/oxxo to go to processing on upe checkout #2066

leonardola committed:
fixed identation #2066

leonardola committed:
fixed identation #2066

leonardola committed:
Merge branch 'develop' into add/2066-add-boleto-and-oxxo

leonardola committed:
Merge remote-tracking branch 'origin/add/2066-add-boleto-and-oxxo' into add/2066-add-boleto-and-oxxo

leonardola commented:
@reykjalin I created the issues for the bugs I found. I've also updated the branch with develop and did a small bug fix with 8a935693f48e385948e4313532b1a5f7ba00ae89 can you please take a look at it and if everything is ok approve so I can merge this?

reykjalin commented on includes/abstracts/abstract-wc-stripe-payment-gateway-voucher.php:
```diff
@@ -0,0 +1,408 @@
+	 *
+	 * @return mixed
+	 */
+	public function add_allowed_payment_processing_statuses( $allowed_statuses, $order ) {
+		$is_voucher_payment = in_array( $order->get_meta( '_stripe_upe_payment_type' ), [ 'boleto', 'oxxo' ] );
+
+		if ( $is_voucher_payment && ! in_array( 'on-hold', $allowed_statuses ) ) {
+			$allowed_statuses[] = 'on-hold';
+		}
+
+		return $allowed_statuses;
+	}
```
I think this should not be in the generic `Voucher` class, and instead added in any classes that extend it 🙂 

So basically; the OXXO class should be responsible for its own allowed payment processing statuses, and the Boleto class should be responsible for its own allowed statuses.

I understand this is a tempting shortcut in this case since the handling is the same for both Boleto and OXXO, but this will look weird if we ever add another voucher-based payment method.

Let's move this check into the classes that extend this one 🙂

reykjalin reviewed (approved):
I have one minor comment, but otherwise this is looking great 🎉 
Excellent work! 👏 

I'm going to preemptively approve the PR, but let's make sure we address the remaining discussion before we merge 👍 

Also, please make sure you add testing instructions for GlobalStep to the [5.9.0 release testing instructions page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-testing-instructions-for-the-WooCommerce-Stripe-payment-gateway-5.9.0). Please make sure this includes extensive testing; so step-by-step testing instructions based on the list you wrote in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2115#issuecomment-977206080.

leonardola committed:
changed how the on-hold is added to the array of valid statuses for webhook processing #2066

areoimiles commented:
Hi,

This change caused our Stripe integration to stop working. Customers weren't able to enter payment details due to a js error, which has resulted in no orders being placed since Thursday. The error that occurs is:

**i("#stripe_boleto_tax_id").mask(n, e) is not a function** which occurs at line 279 in this file **woocommerce-gateway-stripe/assets/js/stripe.js**

To resolve this I have commented out that line and disabled auto updates on the plugin. I'm guessing because we don't have Boleto as a payment option this element doesn't exist and therefor the mask function can't run and breaks the implementation.

Is it worth checking to see if the element exists before running the function:

if ($("#stripe_boleto_tax_id").length) {
     i("#stripe_boleto_tax_id").mask(n, e)
}

dechov commented on includes/class-wc-stripe-intent-controller.php:
```diff
@@ -404,14 +404,16 @@ public function update_payment_intent( $payment_intent_id = '', $order_id = null
+				'amount'      => WC_Stripe_Helper::get_stripe_amount( $amount, strtolower( $currency ) ),
+				'currency'    => strtolower( $currency ),
+				'metadata'    => $gateway->get_metadata_from_order( $order ),
+				'description' => __( 'Stripe - Order', 'woocommerce-gateway-stripe' ) . ' ' . $order->get_id(),
```
Looks like the shortcode flow uses the site title (instead of "Stripe") and the order number (instead of order ID) [[source](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/ace1b9f93a950a62d20d194ce44a74286e99ed2b/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L421)] – should it be the same here? Could open an issue for that if so
